ignition/src/ECMSingleton.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2019 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#include "gympp/gazebo/ECMSingleton.h"
+#include "gympp/Log.h"
+
+#include <mutex>
+#include <ostream>
+#include <unordered_map>
+
+using namespace gympp::gazebo;
+using RobotName = std::string;
+
+class ECMSingleton::Impl
+{
+public:
+    mutable std::mutex mutex;
+    ignition::gazebo::EventManager* eventMgr = nullptr;
+    ignition::gazebo::EntityComponentManager* ecm = nullptr;
+};
+
+ECMSingleton::ECMSingleton()
+    : pImpl{new Impl(), [](Impl* impl) { delete impl; }}
```

Comment:
This seems to be exactly the default deleter, there is any reason to pass it explicitly?

Reply:
I remember I wrote the reason behind this workaround somewhere (maybe blockfactory / wb-toolbox) but I cannot find it anymore. It is a common pattern I typically use.

If I recall, the problem is due to the use of a unique pointer for the Impl object. The header has a forward declaration of the class, and the implementation is in the cpp file. In order to know how to free the memory for the Impl object (handled by the unique pointer), the forward declaration is not enough.

The workaround is initializing the unique pointer with a custom deleter (so you tell the compiler that you will handle it properly somehow), and you specify the deleter in the cpp.

See as an example:

https://github.com/robotology/gym-ignition/blob/aa91b25a23c44305b9a62de74a3c0f7eeeed4006/ignition/include/gympp/gazebo/RobotSingleton.h#L28-L29

Reply:
Ah, I see, I guess the reason is explained in https://www.fluentcpp.com/2017/09/22/make-pimpl-using-unique_ptr/ (where they use another technique to workaround this issue).

Reply:
I remember I read and tried that solution, but still there were cases where it didn't work. I found that doing that plus the custom deleter is a consistent solution that always works (and it does not add any extra line of code).